lib/posix/pthread.c

```diff
@@ -215,20 +215,22 @@ int pthread_create(pthread_t *newthread, const pthread_attr_t *attr,
  */
 int pthread_setcancelstate(int state, int *oldstate)
 {
+	bool cancel_pending;
+	k_spinlock_key_t cancel_key;
 	struct posix_thread *pthread = to_posix_thread(pthread_self());
 
 	if (state != PTHREAD_CANCEL_ENABLE &&
 	    state != PTHREAD_CANCEL_DISABLE) {
 		return EINVAL;
 	}
 
+	cancel_key = k_spin_lock(&pthread->cancel_lock);
 	*oldstate = pthread->cancel_state;
-
-	pthread_mutex_lock(&pthread->cancel_lock);
 	pthread->cancel_state = state;
-	pthread_mutex_unlock(&pthread->cancel_lock);
+	pthread->cancel_pending = cancel_pending;
```

Comment:
This appears to be backwards? Should it be `cancel_pending = pthread->cancel_pending` instead?